### ISSUE TITLE
🚀🚀 Real Time Object Detection in TensorLayer

### ISSUE THREAD

**Issue by zsdonghao** [tensorlayer/tensorlayer#435]
A discussion for real-time object detection using TensorLayer and TensorFlow

- For Model Acceleration, see https://github.com/tensorlayer/tensorlayer/issues/416
- Mobilenet YOLO, see https://github.com/eric612/MobileNet-YOLO/issues/7#issuecomment-416430694

### New Paper list
* [Towards High Performance Video Object Detection for Mobiles](https://arxiv.org/pdf/1804.05830.pdf)  25FPS on Huawei Mate8

### Paper list
* PVANET [[Paper]](https://arxiv.org/pdf/1608.08021) [[Code]](https://github.com/sanghoon/pva-faster-rcnn)
* OverFeat, NYU [[Paper]](http://arxiv.org/pdf/1312.6229.pdf)
* R-CNN, UC Berkeley [[Paper-CVPR14]](http://www.cv-foundation.org/openaccess/content_cvpr_2014/papers/Girshick_Rich_Feature_Hierarchies_2014_CVPR_paper.pdf) [[Paper-arXiv14]](http://arxiv.org/pdf/1311.2524)
* SPP, Microsoft Research [[Paper]](http://arxiv.org/pdf/1406.4729) 
* Fast R-CNN, Microsoft Research [[Paper]](http://arxiv.org/pdf/1504.08083)
* Faster R-CNN, Microsoft Research [[Paper]](http://arxiv.org/pdf/1506.01497)
* R-CNN minus R, Oxford [[Paper]](http://arxiv.org/pdf/1506.06981)
* End-to-end people detection in crowded scenes [[Paper]](http://arxiv.org/abs/1506.04878)
* Yolo1: Unified, Real-Time Object Detection [[Paper]](http://arxiv.org/abs/1506.02640), [[Paper Version 2]](https://arxiv.org/abs/1612.08242), [[C Code]](https://github.com/pjreddie/darknet), [[Tensorflow Code]](https://github.com/thtrieu/darkflow)
* 🚀 Yolo2: [[Paper]](https://arxiv.org/abs/1612.08242), [[C Code]](https://github.com/pjreddie/darknet)
* 🚀 Yolo3: An Incremental Improvement [paper](https://pjreddie.com/media/files/papers/YOLOv3.pdf)
* Inside-Outside Net [[Paper]](http://arxiv.org/abs/1512.04143)
* Deep Residual Network (Current State-of-the-Art) [[Paper]](http://arxiv.org/abs/1512.03385)
* Weakly Supervised Object Localization with Multi-fold Multiple Instance Learning [[Paper](http://arxiv.org/pdf/1503.00949.pdf)]
* R-FCN [[Paper]](https://arxiv.org/abs/1605.06409) [[Code]](https://github.com/daijifeng001/R-FCN)
* 🚀 SSD [[Paper]](https://arxiv.org/pdf/1512.02325v2.pdf) [[Code]](https://github.com/weiliu89/caffe/tree/ssd)
* RefineDet : Single-Shot Refinement Neural Network for Object Detection [[Paper](https://arxiv.org/pdf/1711.06897.pdf)][[Code](https://github.com/sfzhang15/RefineDet)]
* Speed/accuracy trade-offs for modern convolutional object detectors [[Paper]](https://arxiv.org/pdf/1611.10012v1.pdf)
* Focal Loss for Dense Object Detection (RetinaNet) [paper](https://arxiv.org/abs/1708.02002)

### Datasets
Popular datasets for object detection.

* MSCOCO [[link]](http://cocodataset.org/#home)
* VOC2007/2012 [[2007]](http://host.robots.ox.ac.uk/pascal/VOC/voc2007/) [[2012]](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/) : 20 classes, and 3 extra classes for person.
* MPII [[link]](http://human-pose.mpi-inf.mpg.de/#overview)
* ImageNet

### Existing code
- load VOC dataset using TensorLayer [tl.files.load_voc_dataset](http://tensorlayer.readthedocs.io/en/latest/modules/files.html#voc-2007-2012)
- VOC data augmentation using TensorLayer [tl.prepro.obj_box](http://tensorlayer.readthedocs.io/en/latest/modules/prepro.html#object-detection) ,  [知乎](https://zhuanlan.zhihu.com/p/31466173)
- 🚀 VOC data augmentation using TensorLayer and TensorFlow Dataset API, [here](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_tf_dataset_voc.py)
- [darkflow](https://github.com/thtrieu/darkflow) incorrect implementation for YOLO2, don't use!
- [YAD2K: Yet Another Darknet 2 Keras](https://github.com/allanzelener/YAD2K) correct implementation.
- [experiencor/keras-yolo3](https://github.com/experiencor/keras-yolo3)
- [SpikeKing/keras-yolo3-detection](https://github.com/SpikeKing/keras-yolo3-detection/)

### WIP
- YOLO2 (ignore YOLO1) for VOC/ COCO
- SSD for VOC/ COCO
- API to load MSCOCO dataset
- NMS and Soft-NMS
- API for evaluation - mAP
- https://github.com/tensorlayer/tensorlayer/issues/371

### Others
- [TensorFlow lite](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/lite)
- [Object Localization and Detection](https://leonardoaraujosantos.gitbooks.io/artificial-inteligence/content/object_localization_and_detection.html)
- [Soft NMS](https://github.com/bharatsingh430/soft-nms/blob/master/lib/nms/cpu_nms.pyx) - can get 1mAP improve when using non one short method.
- [mAP](https://stackoverflow.com/questions/36274638/map-metric-in-object-detection-and-computer-vision)
- Use TF dataset API to deal with large dataset https://blog.csdn.net/dQCFKyQDXYm3F8rB0/article/details/79342369
- [Receptive Fields Calculator](https://fomoro.com/tools/receptive-fields/#3,1,1,SAME;3,1,1,SAME;2,2,1,SAME;3,1,1,SAME;3,1,1,SAME;2,2,1,SAME;3,1,1,SAME;3,1,1,SAME;2,2,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME;3,1,1,SAME)

**Comment by auroua** [comment]
I think `RefineDet` is also a good object detection model recently[[Paper](https://arxiv.org/pdf/1711.06897.pdf)][[Code](https://github.com/sfzhang15/RefineDet)]

**Comment by auroua** [comment]
I think `RefineDet` is also a good object detection model recently[[Paper](https://arxiv.org/pdf/1711.06897.pdf)][[Code](https://github.com/sfzhang15/RefineDet)]

**Comment by zsdonghao** [comment]
@auroua added to the paper list

**Comment by hanm** [comment]
another paper :  [Focal Loss for Dense Object Detection (RetinaNet)](https://arxiv.org/abs/1708.02002)

**Comment by zsdonghao** [comment]
I release example code for using TensorFlow Dataset API to preprocess VOC data [here](https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_tf_dataset_voc.py)
